### PR TITLE
refactor(activation_based): make memory modules functional-backed

### PR DESCRIPTION
**Pull request by AllenYolk**
## Summary

- Add native grouped functional APIs to `MemoryModule`: inputs and outputs precede explicit state tuples, with separate single-step and multi-step implementations.
- Back regular `single_step_forward` and `multi_step_forward` with the functional path and centralize scalar-to-tensor state initialization in `materialize_states`.
- Keep `to_functional_forward` as the public conversion entry point, selecting native functional execution when available and the state-swap fallback for composite or Simple modules.
- Migrate forward-capable production `MemoryModule` implementations across neurons, layers, encoders, ANN2SNN recipes, containers, checkpointing, and related tooling.
- Keep charge-fire-reset customization on `SimpleBaseNode` / `SimpleIFNode` / `SimpleLIFNode`; production neurons such as `LIFNode` now own a sealed functional equation.

## Breaking changes

- Production `BaseNode` neurons no longer expose `neuronal_charge`, `neuronal_fire`, or `neuronal_reset` as customization hooks. Users overriding those equations should migrate to the `Simple*Node` family or implement the functional forward contract.
- Calling a `MemoryModule` subclass without forward semantics now raises through the base functional method. `STDPLearner`, `MSTDPLearner`, and `MSTDPETLearner` remain state-lifecycle helpers whose computation entry point is `step()`.
- The migration and fallback behavior are documented in the bilingual neuron and memory-optimization tutorials and in the changelog.

## Audit and review

A runtime audit covered 72 framework `MemoryModule` subclasses, excluding optional CUDA/Triton and example packages. All forward-capable production modules now have a native functional path. Intentional exceptions are:

- three `Simple*Node` classes using charge-fire-reset plus `to_functional_forward` fallback;
- three `step()`-only learning helpers;
- conceptual base classes and three native multi-step-only neurons (`DSRIFNode`, `DSRLIFNode`, `GatedLIFNode`).

Independent spec and standards reviews found and prompted fixes for the RL readout examples, repeated state materialization, operation-counter state identification, stale legacy links, and incomplete concrete type/docs details. A suggested aggregate op-counter assertion was not restored because it encoded the old hidden execution graph rather than the public state-access contract; the state-read regression remains protected.

## Validation

- `PYTHONPATH=. uv run pytest -q test/activation_based`: **1620 passed, 481 skipped**
- `uv run ruff format --check <changed Python files>`: passed
- `uv run ruff check <changed Python files>`: passed
- `uv run python tools/generate_changelog_rst.py --check`: passed
- `uv run sphinx-build -M html docs/source docs/build -E`: passed with 0 warnings
- `git diff --check`: passed

A same-environment CPU latency check (`torch.set_num_threads(1)`, gradients disabled, blocked autorange) compared this branch with `origin/master`:

| Node/path | master | this branch | median latency change |
| --- | ---: | ---: | ---: |
| IF single-step | 33.039 us | 31.453 us | -4.8% |
| IF multi-step | 552.720 us | 499.569 us | -9.6% |
| LIF single-step | 37.402 us | 35.441 us | -5.2% |
| LIF multi-step | 619.800 us | 584.542 us | -5.7% |

These are local CPU forward-latency measurements, not a universal speedup claim.

## Summary by CodeRabbit

* **New Features**
  * Added explicit-state functional forwarding across neurons, encoders, layers, recurrent modules, attention, and ANN-to-SNN components.
  * Added automatic state materialization for input shape, device, and data type compatibility.
  * Improved single-step and multi-step execution, composition, backend support, and functional neuron utilities.

* **Breaking Changes**
  * Functional forwarding now uses grouped inputs and states, returning outputs with updated states.
  * Removed legacy neuron transition hooks and `v_float_to_tensor()`.
  * Unsupported module calls now raise `NotImplementedError`.

* **Documentation**
  * Updated tutorials, API references, and changelogs with migration guidance.

### CODE REVIEW

File: docs/source/changelog.rst [hunk 1]
```diff
@@ -51,6 +52,24 @@ Module: ``spikingjelly.activation_based.functional``.
...
+  modules use a dictionary-swap fallback. Custom production ``MemoryModule`` and
+  ``BaseNode`` subclasses should implement ``single_step_functional_forward`` and
+  override ``multi_step_functional_forward`` only when they provide a specialized
+  sequence implementation. ``SimpleBaseNode``, ``SimpleIFNode``, and ``SimpleLIFNode``
+  retain the charge-fire-reset forward model for equation experiments and use the
+  general state-substitution path when converted.
+- Added ``MemoryModule.materialize_states(inputs, states, step_mode)`` as the
+  single hook for converting registered scalar or empty states into
+  input-dependent states before functional execution. Its default implementation
+  returns states unchanged.
```
Comment (AllenYolk) [inline docs/source/changelog.rst]: Too verbose. Make it concise

File: docs/source/changelog.rst [hunk 2]
```diff
@@ -227,6 +246,43 @@ Module: ``spikingjelly.activation_based.triton_kernel.neuron_kernel``.
...
+  ``LIFNode`` use native functional transitions and their specialized sequence
+  kernels.
+- **Breaking change:** removed ``BaseNode.v_float_to_tensor()``. Custom functional
+  modules that require input-dependent state initialization should override
+  ``materialize_states(inputs, states, step_mode)`` and return a new state tuple
+  without mutating module memory.
+- **Breaking change:** ``MemoryModule`` subclasses without functional or regular
+  forward semantics now raise ``NotImplementedError`` when called. Stateful
+  learners such as ``STDPLearner``, ``MSTDPLearner``, and ``MSTDPETLearner`` continue
+  to expose their computation through ``step()``.
```
Comment (AllenYolk) [inline docs/source/changelog.rst]: Too verbose. Make it concise

File: docs/source/changelog.rst [hunk 3]
```diff
@@ -227,6 +246,43 @@ Module: ``spikingjelly.activation_based.triton_kernel.neuron_kernel``.
...
+
+Module: ``spikingjelly.activation_based.base``.
+
+- **Breaking change:** ``to_functional_forward()`` now returns a function with the
+  grouped interface
+  ``(inputs, states, **kwargs) -> (outputs, updated_states)``. Both ``inputs`` and
+  ``outputs`` remain tuples when they contain one tensor. Replace calls such as
+  ``output, state = forward(x, state)`` with
+  ``outputs, states = forward((x,), (state,))`` and read the single output from
+  ``outputs[0]``.
```
Comment (AllenYolk) [inline docs/source/changelog.rst:235]: Too verbose. Make it concise.

File: docs/source/legacy_tutorials/cn/0_neuron.rst [hunk 4]
```diff
@@ -98,7 +98,11 @@ LIF神经元层有一些构造参数，在API文档中对这些参数有详细
             else:
                 self.v += (x - (self.v - self.v_reset)) / self.tau
 
-不同的神经元，充电方程不尽相同。但膜电位超过阈值电压后，释放脉冲，以及释放脉冲后，膜电位的重置都是相同的。因此它们全部继承自 :class:`spikingjelly.activation_based.neuron.BaseNode`，共享相同的放电、重置方程。可以在 :class:`spikingjelly.activation_based.neuron.BaseNode.neuronal_fire` 中找到释放脉冲的代码：
+不同的神经元具有不同的充电方程。:class:`spikingjelly.activation_based.neuron.SimpleBaseNode`
+是为了教学和动力学自定义而提供的纯 PyTorch 接口，并不是一种神经元数学模型。它直接
+展示神经元的充电、放电和重置职责；继承该接口的实现共享放电和重置方程。
```
Comment (AllenYolk) [inline docs/source/legacy_tutorials/cn/0_neuron.rst]: Don't use "negative tone" in docs or tutorials!
Firing and reset equations can be overwritten. They are not shared.

File: docs/source/legacy_tutorials/cn/0_neuron.rst [hunk 5]
```diff
@@ -113,7 +117,9 @@ LIF神经元层有一些构造参数，在API文档中对这些参数有详细
 
 #. Soft方式：释放脉冲后，膜电位减去阈值电压：:math:`V[t] = V[t] - V_{threshold}`
 
-可以发现，对于使用Soft方式的神经元，并不需要重置电压 :math:`V_{reset}` 这个变量。:class:`spikingjelly.activation_based.neuron` 中的神经元，在构造函数的参数之一 ``v_reset``，默认为 ``1.0`` ，表示神经元使用Hard方式；若设置为 ``None``，则会使用Soft方式。在 :class:`spikingjelly.activation_based.neuron.BaseNode.neuronal_fire.neuronal_reset` 中可以找到膜电位重置的代码：
+可以发现，对于使用Soft方式的神经元，并不需要重置电压 :math:`V_{reset}` 这个变量。:class:`spikingjelly.activation_based.neuron` 中的神经元，在构造函数的参数之一 ``v_reset``，默认为 ``1.0`` ，表示神经元使用Hard方式；若设置为 ``None``，则会使用Soft方式。在
```
Comment (AllenYolk) [inline docs/source/legacy_tutorials/cn/0_neuron.rst]: `v_reset` defaults to `0.0`.

File: docs/source/tutorials/cn/memopt.rst [hunk 6]
```diff
@@ -371,6 +371,24 @@ Step 1. 定义分割规则
...
+
+    f_forward = base.to_functional_forward(neuron.LIFNode())
+    outputs, updated_states = f_forward((x,), (v,))
+
+已实现 functional forward 的 MemoryModule 会直接执行显式状态转移；
+``nn.Sequential`` 会递归组合子模块；其他复合模块由通用路径处理。
+``SimpleBaseNode``、``SimpleIFNode`` 和 ``SimpleLIFNode`` 的名称表示其纯 PyTorch、
+便于理解和自定义动力学的接口定位，并不表示一种神经元数学模型。它们为保留可修改的
+``charge → fire → reset`` 方程接口，也使用通用路径：转换函数会临时换入显式状态，
+执行原有前向后恢复模块状态。该路径保持数值与状态语义，但包含状态替换开销。
```
Comment (AllenYolk) [inline docs/source/tutorials/cn/memopt.rst]: No need to mention the simple nodes here.

File: spikingjelly/activation_based/ann2snn/recipes/sta_transformer.py [hunk 7]
```diff
@@ -198,40 +198,62 @@ def __init__(
...
+        if step_mode == "m":
+            if x.dim() < 2:
+                raise ValueError(
+                    "STA spike encoder multi-step forward expects an input "
+                    "sequence with a time dimension and at least one data dimension."
+                )
+            if x.shape[0] == 0:
+                raise ValueError(
+                    "STA spike encoder expects a non-empty time dimension."
+                )
```
Comment (AllenYolk) [inline spikingjelly/activation_based/ann2snn/recipes/sta_transformer.py]: Not needed. Remove the guard.

File: spikingjelly/activation_based/ann2snn/recipes/sta_transformer.py [hunk 8]
```diff
@@ -198,40 +198,62 @@ def __init__(
...
-        for x in x_seq:
-            outputs.append(self.single_step_forward(x))
-        return torch.stack(outputs, dim=0)
+    def multi_step_functional_forward(
+        self,
+        inputs: tuple[torch.Tensor, ...],
+        states: tuple[object, ...],
+        **kwargs: object,
+    ) -> tuple[tuple[torch.Tensor, ...], tuple[object, ...]]:
+        return super().multi_step_functional_forward(inputs, states, **kwargs)
```
Comment (AllenYolk) [inline spikingjelly/activation_based/ann2snn/recipes/sta_transformer.py]: No need to explicitly define it?

File: spikingjelly/activation_based/ann2snn/recipes/transformer_td_equivalent.py [hunk 9]
```diff
@@ -44,8 +44,13 @@ class _TDTanh(TDModule):
     def ann_forward(self, x: torch.Tensor) -> torch.Tensor:
         return torch.tanh(x)
 
-    def multi_step_forward(self, x_seq: torch.Tensor) -> torch.Tensor:
-        return self._td_sequence_forward((x_seq,), torch.tanh)
+    def multi_step_functional_forward(
+        self,
+        inputs: tuple[torch.Tensor, ...],
+        states: tuple[object, ...],
```
Comment (AllenYolk) [inline spikingjelly/activation_based/ann2snn/recipes/transformer_td_equivalent.py]: Use `Any` instead of `object` for type hint

File: spikingjelly/activation_based/functional/neuron.py [hunk 10]
```diff
@@ -815,11 +815,16 @@ def lif_step(
        :func:`lif_multi_step_inductor`, :func:`lif_multi_step_cupy`,
        :func:`lif_multi_step_triton`.
     """
-    v_reset_value = 0.0 if v_reset is None else v_reset
     if decay_input:
-        v_charged = v + (x - (v - v_reset_value)) / tau
+        if v_reset is None or v_reset == 0.0:
+            v_charged = v + (x - v) / tau
+        else:
+            v_charged = v + (x - (v - v_reset)) / tau
```
Comment (AllenYolk) [inline spikingjelly/activation_based/functional/neuron.py]: No need to separate `v_reset == 0.0` from the general case. For pytorch backend, simplicity is more important than performance.

File: spikingjelly/activation_based/neuron/flexsn.py [hunk 11]
```diff
@@ -1315,42 +1315,67 @@ def init_states(num_states: int, step_mode: str, *args) -> List[torch.Tensor]:
...
-
-    def multi_step_forward(self, *args):
-        T = args[0].shape[0]
+    def single_step_functional_forward(
+        self,
+        inputs: tuple[torch.Tensor, ...],
+        states: tuple[object, ...],
+        **kwargs: object,
+    ) -> tuple[tuple[torch.Tensor, ...], tuple[object, ...]]:
+        state_values = states[1] if self.store_state_seqs else states[0]
```
Comment (AllenYolk) [inline spikingjelly/activation_based/neuron/flexsn.py]: **How is 'states' organized???**

functional interface should not handle store_state_seqs.

**Review by AllenYolk — COMMENTED** [review]
LGTM
